[FSDP] Fix 2D fully sharded state_dict regression.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103214
* #103172

The #100856 PR changed how we load a fully sharded state dict.

Originally, we'd pick the unsharded params from the state_dict,
apply 2D transform and gather them. Then we'd use those tensors
to create a flat param tensor and put that in the state_dict.

The code in module::load_state_dict would then load the flat param
just fine.

Then #100856 changed to keep the state_dict with the unsharded params and
unshard the model as well.

The problem is that the extension peeled out the nested sharded tensor but
was not given an opportunity to unpeel back into a DTensor.

The fix is to just do that, convert the gathered tensor into the DTensor
shard it actually is.

Fixes #101162

kumpera committed:
[FSDP] Fix 2D fully sharded state_dict regression.

The #100856 PR changed how we load a fully sharded state dict.

Originally, we'd pick the unsharded params from the state_dict,
apply 2D transform and gather them. Then we'd use those tensors
to create a flat param tensor and put that in the state_dict.

The code in module::load_state_dict would then load the flat param
just fine.

Then #100856 changed to keep the state_dict with the unsharded params and
unshard the model as well.

The problem is that the extension peeled out the nested sharded tensor but
was not given an opportunity to unpeel back into a DTensor.

The fix is to just do that, convert the gathered tensor into the DTensor
shard it actually is.

Fixes #101162

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -634,6 +638,10 @@ def _sharded_pre_load_state_dict_hook(
                 tensor, local_tensor, group=fsdp_state.process_group
             )
         tensor = tensor.narrow(0, 0, param_numel).reshape(param.size())
+        if fqn_to_param_ext.get(fqn) is not None:
```
Curious: When do we expect `fqn_to_param_ext.get(fqn)` to not have `fqn` in `fqn_to_param_ext` and hence return the default `get()` value of `None` (i.e. not because the element in `_param_extenssions` is `None`)?

fegin replied:
Yup, I think that's when only FSDP is applied, not 2D's case

kumpera replied:
This happens when the FQN is for a regular tensor and not a DTensor.

awgu commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -586,10 +590,10 @@ def _sharded_pre_load_state_dict_hook(
-        fqn for fqn, _, _ in _shared_param_name_infos(module, fsdp_state)
-    ]
+    fqn_to_param_ext = {
+        t[0]: t[1]
+        for t in zip(handle.flat_param._fqns, handle.flat_param._param_extensions)
```
nit: maybe for readability?
```suggestion
        fqn: param_extension
        for fqn, param_extension in zip(handle.flat_param._fqns, handle.flat_param._param_extensions)
```

awgu commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -634,6 +638,10 @@ def _sharded_pre_load_state_dict_hook(
             )
         tensor = tensor.narrow(0, 0, param_numel).reshape(param.size())
+        if fqn_to_param_ext.get(fqn) is not None:
+            ext = fqn_to_param_ext.get(fqn)
```
I wonder if we are allowed to use the walrus operator now.

kumpera replied:
Good question, we should since it's on python 3.8

awgu commented:
I will let @fegin or @wz337 approve. I was mainly just reading the PR for my own understanding.

fegin reviewed (commented):
Not about this PR, but would it be better to use state_dict_hook and load_state_dict_hook to implement the conversion for 2D? I feel hook-based implementation will be a cleaner implementation and post less composability issues. It won't have the case where FSDP change the code and does not know where and when to call the extension.

fegin reviewed (approved):
Accept to unblock. @awgu's comment is valid. Please change it accordingly.

fduwjj reviewed (commented):
N00b question, do we have unit test to capture this failure in future?

kumpera committed:
Update on "[FSDP] Fix 2D fully sharded state_dict regression."


The #100856 PR changed how we load a fully sharded state dict.

Originally, we'd pick the unsharded params from the state_dict,
apply 2D transform and gather them. Then we'd use those tensors
to create a flat param tensor and put that in the state_dict.

The code in module::load_state_dict would then load the flat param
just fine.

Then #100856 changed to keep the state_dict with the unsharded params and
unshard the model as well.

The problem is that the extension peeled out the nested sharded tensor but
was not given an opportunity to unpeel back into a DTensor.

The fix is to just do that, convert the gathered tensor into the DTensor
shard it actually is.

Fixes #101162

[ghstack-poisoned]

kumpera committed:
Update on "[FSDP] Fix 2D fully sharded state_dict regression."


The #100856 PR changed how we load a fully sharded state dict.

Originally, we'd pick the unsharded params from the state_dict,
apply 2D transform and gather them. Then we'd use those tensors
to create a flat param tensor and put that in the state_dict.

The code in module::load_state_dict would then load the flat param
just fine.

Then #100856 changed to keep the state_dict with the unsharded params and
unshard the model as well.

The problem is that the extension peeled out the nested sharded tensor but
was not given an opportunity to unpeel back into a DTensor.

The fix is to just do that, convert the gathered tensor into the DTensor
shard it actually is.

Fixes #101162

[ghstack-poisoned]

kumpera committed:
Update on "[FSDP] Fix 2D fully sharded state_dict regression."


The #100856 PR changed how we load a fully sharded state dict.

Originally, we'd pick the unsharded params from the state_dict,
apply 2D transform and gather them. Then we'd use those tensors
to create a flat param tensor and put that in the state_dict.

The code in module::load_state_dict would then load the flat param
just fine.

Then #100856 changed to keep the state_dict with the unsharded params and
unshard the model as well.

The problem is that the extension peeled out the nested sharded tensor but
was not given an opportunity to unpeel back into a DTensor.

The fix is to just do that, convert the gathered tensor into the DTensor
shard it actually is.

Fixes #101162

[ghstack-poisoned]

wz337 commented:
> N00b question, do we have unit test to capture this failure in future?

Both test_2d_parallel(https://github.com/pytorch/pytorch/blob/4aee0fef117df1975200984a0e6c36eb8a078bf2/test/distributed/tensor/parallel/test_2d_parallel.py#L108) and test_2d_fsdp_dt_checkpoint.py(https://github.com/pytorch/pytorch/blob/main/test/distributed/checkpoint/test_2d_fsdp_dt_checkpoint.py) should be able to capture this.
